KAFKA-3517: Add documentation for SASL/PLAIN

rajinisivaram opened the pull request:
Documentation corresponding to KIP-43 - SASL/PLAIN and multiple mechanism support.

rajinisivaram committed:
KAFKA-3517: Add documentation for SASL/PLAIN

rajinisivaram commented:
The code changes corresponding to this doc are under KAFKA-3149 (not yet committed).

rajinisivaram committed:
KAFKA-3517: Add broker config sasl.mechanism.inter.broker.protocol

edenhill commented on docs/security.html:
```diff
@@ -257,93 +321,191 @@
+        user_admin="admin-secret"
+        user_alice="alice-secret";
+    };</pre></li>
+          This configuration defines two users (<i>admin</i> and <i>alice</i>). <i>admin</i> is used for inter-broker communication.
```
Maybe explicitly mention that alice is a client user?
Is user_N defining a client user? If so, does this mean that admin may also connect as a client in the above example?
I'm guessing username/password is for outgoing sessions, while user_N is for incoming?

rajinisivaram replied:
@edenhill Thank you for the review. `username` and `password` define the username/password combination used by client connections. Kafka clients use the login context `KafkaClient` and brokers use the login context `KafkaServer`. So in the example, the client connections use the username `alice` and the inter-broker connections use the username `admin`. In KafkaServer, `user_N=M` is the server-side configuration defining all the valid username/password pairs that the server validates against. In this example, any client may connect using username `admin` by setting `username=admin` in the `KafkaClient` context. Hope that helps. I will update the docs to clarify these.

junrao replied:
Yes, second on Magnus question. We should make it clear that username/password is used for broker to initiate a connection to other brokers and user_userName is used for broker to validate all client connections (including those from other brokers).

rajinisivaram replied:
Have updated the docs.

junrao commented on docs/security.html:
```diff
@@ -182,22 +183,95 @@
+  </li>
+  <li><h4><a id="security_sasl_clientconfig"
+    href="#security_sasl_clientconfig">SASL configuration for Kafka clients</a></h4>
+    SASL authentication is only supported for the new kafka producer and
```
To make it clear, perhaps change "new" to "new java".

rajinisivaram replied:
Done.

junrao commented on docs/security.html:
```diff
@@ -257,93 +322,191 @@
+      </li>
+
+      <li><h5><a id="security_sasl_plain_clientconfig" href="#security_sasl_plain_clientconfig">Configuring Kafka Clients</a></h5>
+        To configure SASL authentication  on the clients:
```
extra space before on.

rajinisivaram replied:
Removed.

junrao commented on docs/security.html:
```diff
@@ -257,93 +322,191 @@
+            <code>org.apache.kafka.common.security.plain.PlainLoginModule</code> can be used as an example.</li>
+          <li>In production systems, external authentication servers may implement password authentication. Kafka brokers can be integrated with these servers by adding
+            your own implementation of <code>javax.security.sasl.SaslServer</code>. The default implementation included in Kafka in the package
+            <code>org.apache.kafka.common.security.plain</code> can be used as an example to get started.</li>
```
Could we describe how the user can plug in a customized PLAIN implementation (e.g., configuring the provider)?

rajinisivaram replied:
Done.

junrao commented:
@rajinisivaram : Thanks for the patch. Could you also add a description of the latest SASL authentication protocol in protocol.html?

rajinisivaram committed:
KAFKA-3517: Address review comments, update protocol docs.

rajinisivaram commented:
@edenhill @junrao Thank you both for the reviews. I have updated the docs and added a section in protocol.html. Please let me know if more details should be added.

junrao commented:
@rajinisivaram : Thanks for the patch. LGTM
